[RFC] Extend #[\Override] to target properties

jiripudil opened the pull request:
https://wiki.php.net/rfc/override_properties

TimWolla commented on Zend/tests/attributes/override/properties_20.phpt:
```diff
@@ -0,0 +1,16 @@
+}
+
+echo "Done";
+
```
Please add the closing PHP tag to all tests.

DanielEScherzer commented on Zend/tests/attributes/override/properties_02.phpt:
```diff
@@ -0,0 +1,16 @@
+--FILE--
+<?php
+
+class C
+{
```
```suggestion
class C {
```

to match the other tests

DanielEScherzer commented on Zend/tests/attributes/override/properties_07.phpt:
```diff
@@ -0,0 +1,19 @@
+--TEST--
+#[\Override]: Properties: On used trait without parent method.
```
```suggestion
#[\Override]: Properties: On used trait without parent property.
```

DanielEScherzer commented on Zend/tests/attributes/override/properties_08.phpt:
```diff
@@ -0,0 +1,23 @@
+--TEST--
+#[\Override]: Properties: On used trait with interface method.
```
```suggestion
#[\Override]: Properties: On used trait with interface property.
```

DanielEScherzer commented on Zend/tests/attributes/override/properties_10.phpt:
```diff
@@ -0,0 +1,19 @@
+--TEST--
+#[\Override]: Properties: Parent property is private (2).
```
For the tests where this is a (2) version, I suggest identifying what the difference is, e.g.
```
#[\Override]: Property: Parent property is private (child property public)
#[\Override]: Property: Parent property is private (child property private)
```

DanielEScherzer commented on Zend/zend_compile.c:
```diff
@@ -7881,6 +7881,11 @@ static void zend_compile_params(zend_ast *ast, zend_ast *return_type_ast, uint32
 					&prop->attributes, attributes_ast, 0, ZEND_ATTRIBUTE_TARGET_PROPERTY, ZEND_ATTRIBUTE_TARGET_PARAMETER);
 			}
+
+			zend_attribute *override_attribute = zend_get_attribute_str(prop->attributes, "override", sizeof("override")-1);
```
Checking for the override attribute should only be needed if the property has any attributes, i.e. this can be moved into the `if (attributes_ast) {` conditional block

DanielEScherzer commented on Zend/zend_compile.c:
```diff
@@ -8859,6 +8864,11 @@ static void zend_compile_prop_decl(zend_ast *ast, zend_ast *type_ast, uint32_t f
 			zend_compile_attributes(&info->attributes, attr_ast, 0, ZEND_ATTRIBUTE_TARGET_PROPERTY, 0);
 		}
 
+		zend_attribute *override_attribute = zend_get_attribute_str(info->attributes, "override", sizeof("override")-1);
```
same as above, check is only needed if the `attr_ast` check passes

TimWolla commented on Zend/zend_inheritance.c:
```diff
@@ -1566,6 +1566,10 @@ static void do_inherit_property(zend_property_info *parent_info, zend_string *ke
 						ZSTR_VAL(parent_info->ce->name));
 			}
+
+			if (child_info->flags & ZEND_ACC_OVERRIDE) {
+				child_info->flags &= ~ZEND_ACC_OVERRIDE;
+			}
```
This can just unconditionally try to remove the flag:

```suggestion
			child_info->flags &= ~ZEND_ACC_OVERRIDE;
```

TimWolla commented on Zend/zend_inheritance.c:
```diff
@@ -2320,6 +2325,15 @@ void zend_inheritance_check_override(const zend_class_entry *ce)
 		}
 	} ZEND_HASH_FOREACH_END();
 
+	ZEND_HASH_MAP_FOREACH_PTR(&ce->properties_info, p) {
+		if (p->flags & ZEND_ACC_OVERRIDE) {
+			zend_error_noreturn(
+				E_COMPILE_ERROR,
+				"%s::$%s has #[\\Override] attribute, but no matching parent property exists",
+				ZSTR_VAL(ce->name), zend_get_unmangled_property_name(p->name));
+		}
+	} ZEND_HASH_FOREACH_END();
```
This can probably merged with the loop below (by removing the `ce->num_hooked_props` check).

jiripudil commented:
Thanks for looking over this. I think I've addressed all the comments, please let me know if there's anything else. The RFC has been approved, so if the PR looks ok as is, I'll rebase the branch and regenerate the conflicting arginfo.

rino1 commented:
when will it be available?

DanielEScherzer commented:
> when will it be available?

It is already available in a testing version as PHP 8.5.0beta1

It will be available in a production version as PHP 8.5.0 in November

DanielEScherzer commented:
> Is it possible to receive a notification? Is there an official page that announces releases?

Yes to both
* new releases are announced on https://news-web.php.net/php.announce (as well as some other channels, but the announcements list is probably the quietest)
* they are also listed on https://php.net/ in a header for a few days
* the whole list of releases is on https://www.php.net/releases/index.php
